sympy/physics/mechanics/tests/test_jointsmethod.py

```diff
@@ -0,0 +1,89 @@
+from sympy import symbols, Matrix, cos, sin, expand
+from sympy.physics.mechanics import (PinJoint, JointsMethod, Body, KanesMethod,
+                                    PrismaticJoint, LagrangesMethod)
+from sympy.physics.vector import dynamicsymbols, ReferenceFrame
+
+
+t = dynamicsymbols._t
+
+
+def test_jointsmethod():
+    P = Body('P')
+    C = Body('C')
+    Pin = PinJoint('P1', P, C)
+    C_ixx, g = symbols('C_ixx g')
+    theta, omega = dynamicsymbols('theta_P1, omega_P1')
+    P.apply_force(g*P.y)
+    method = JointsMethod(P, Pin)
+    assert method.frame == P.frame
+    assert method.bodies == [C, P]
+    assert method.loads == [(P.masscenter, g*P.frame.y)]
+    assert method.q == [theta]
+    assert method.u == [omega]
+    assert method.kdes == [omega - theta.diff()]
+    soln = method.form_eoms()
+    assert soln == Matrix([[-C_ixx*omega.diff()]])
+    assert method.forcing_full == Matrix([[omega], [0]])
+    assert method.mass_matrix_full == Matrix([[1, 0], [0, C_ixx]])
+    assert isinstance(method.method, KanesMethod)
+
+def test_complete_simple_double_pendulum():
+    q1, q2 = dynamicsymbols('q1 q2')
+    u1, u2 = dynamicsymbols('u1 u2')
+    m, l = symbols('m l')
+    N = ReferenceFrame('N')
+    A = ReferenceFrame('A')
+    B = ReferenceFrame('B')
+    C = Body('C', frame=N)  # ceiling
+    PartP = Body('P', frame=A, mass=m)
+    PartR = Body('R', frame=B, mass=m)
+
+    J1 = PinJoint('J1', C, PartP, speeds=u1, coordinates=q1,
+                  child_joint_pos=-l*A.x, parent_axis=C.frame.z,
+                  child_axis=PartP.frame.z)
+    J2 = PinJoint('J2', PartP, PartR, speeds=u2, coordinates=q2,
+                  child_joint_pos=-l*B.x, parent_axis=PartP.frame.z,
+                  child_axis=PartR.frame.z)
+
+    method = JointsMethod(C, J1, J2)
+    method.form_eoms()
+    assert expand(method.mass_matrix_full) == Matrix([[1, 0, 0, 0], [0, 1, 0, 0],
+                                        [0, 0, 2*l**2*m*cos(q2) + 3*l**2*m, l**2*m*cos(q2) + l**2*m],
+                                        [0, 0, l**2*m*cos(q2) + l**2*m, l**2*m]])
+    assert expand(method.forcing_full) == Matrix([[u1], [u2], [2*l**2*m*u1*u2*sin(q2) +
+                                    l**2*m*u2**2*sin(q2)], [-l**2*m*u1**2*sin(q2)]])
```

Comment:
q1 should be present in these equations of motion. Here is an example result: https://github.com/pydy/pydy/blob/master/examples/double_pendulum/double_pendulum.ipynb

Reply:
Check the equations of motion against the known solution for a simple double pendulum: https://en.wikipedia.org/wiki/Double_pendulum

Reply:
Here is the output from the pydy example:

```
Python 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.25.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: %paste
from sympy import symbols
from sympy.physics.mechanics import *

q1, q2 = dynamicsymbols('q1 q2')
q1d, q2d = dynamicsymbols('q1 q2', 1)
u1, u2 = dynamicsymbols('u1 u2')
u1d, u2d = dynamicsymbols('u1 u2', 1)
l, m, g = symbols('l m g')

N = ReferenceFrame('N')
A = N.orientnew('A', 'Axis', [q1, N.z])
B = N.orientnew('B', 'Axis', [q2, N.z])

A.set_ang_vel(N, u1 * N.z)
B.set_ang_vel(N, u2 * N.z)

O = Point('O')
P = O.locatenew('P', l * A.x)
R = P.locatenew('R', l * B.x)

O.set_vel(N, 0)
P.v2pt_theory(O, N, A)
R.v2pt_theory(P, N, B)

ParP = Particle('ParP', P, m)
ParR = Particle('ParR', R, m)

kd = [q1d - u1, q2d - u2]
FL = [(P, m * g * N.x), (R, m * g * N.x)]
BL = [ParP, ParR]


KM = KanesMethod(N, q_ind=[q1, q2], u_ind=[u1, u2], kd_eqs=kd)

## -- End pasted text --

In [2]: fr, frstar = KM.kanes_equations(BL, FL)

In [3]: fr
Out[3]: 
Matrix([
[-2*g*l*m*sin(q1(t))],
[  -g*l*m*sin(q2(t))]])

In [4]: frstar
Out[4]: 
Matrix([
[-l**2*m*(sin(q1(t))*sin(q2(t)) + cos(q1(t))*cos(q2(t)))*Derivative(u2(t), t) + l**2*m*(-sin(q1(t))*cos(q2(t)) + sin(q2(t))*cos(q1(t)))*u2(t)**2 - 2*l**2*m*Derivative(u1(t), t)],
[   -l**2*m*(sin(q1(t))*sin(q2(t)) + cos(q1(t))*cos(q2(t)))*Derivative(u1(t), t) + l**2*m*(sin(q1(t))*cos(q2(t)) - sin(q2(t))*cos(q1(t)))*u1(t)**2 - l**2*m*Derivative(u2(t), t)]])
```

Reply:
This is formulated the same as the wikipedia article though.

Reply:
Here is the problem redone with coordinates defined relative to the prior joint:

```
Python 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.25.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sympy as sm

In [2]: import sympy.physics.mechanics as me

In [3]: q1, q2, u1, u2 = me.dynamicsymbols('q1, q2, u1, u2')

In [4]: l, m, g = sm.symbols('l, m, g')

In [5]: N = me.ReferenceFrame('N')

In [6]: A = me.ReferenceFrame('A')

In [7]: B = me.ReferenceFrame('B')

In [8]: A.orient_axis(N, N.z, q1)

In [9]: B.orient_axis(A, A.z, q2)

In [10]: A.set_ang_vel(N, u1*N.z)

In [11]: B.set_ang_vel(A, u2*A.z)

In [12]: O = me.Point('O')

In [13]: P = O.locatenew('P', l*A.x)

In [14]: R = P.locatenew('R', l*B.x)

In [15]: O.set_vel(N, 0)

In [16]: P.v2pt_theory(O, N, A)
Out[16]: l*u1(t)*A.y

In [17]: R.v2pt_theory(P, N, B)
Out[17]: l*u1(t)*A.y + l*(u1(t) + u2(t))*B.y

In [18]: bob1 = Particle('B1', P, m)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-18-731beaa5f967> in <module>
----> 1 bob1 = Particle('B1', P, m)

NameError: name 'Particle' is not defined

In [19]: bob1 = me.Particle('B1', P, m)

In [20]: bob2 = me.Particle('B2', R, m)

In [21]: kdes = [q1.diff() - u1, q2.diff() - u2]

In [22]: loads = ((P, m*g*N.x), (R, m*g*N.x))

In [23]: bodies = (bob1, bob2)

In [24]: kane = me.KanesMethod(N, q_ind=(q1, q2), u_ind=(u1, u2), kd_eqs=kdes)

In [25]: fr, frstar = kane.kanes_equations(bodies, loads)

In [26]: fr
Out[26]: 
Matrix([
[g*l*m*(-sin(q1(t))*cos(q2(t)) - sin(q2(t))*cos(q1(t))) - 2*g*l*m*sin(q1(t))],
[                     g*l*m*(-sin(q1(t))*cos(q2(t)) - sin(q2(t))*cos(q1(t)))]])

In [27]: frstar
Out[27]: 
Matrix([
[l**2*m*(u1(t) + u2(t))**2*sin(q2(t)) - l**2*m*u1(t)**2*sin(q2(t)) - m*(l**2*cos(q2(t)) + l**2)*Derivative(u2(t), t) - (l**2*m + m*(2*l**2*cos(q2(t)) + 2*l**2))*Derivative(u1(t), t)],
[                                                                         -l**2*m*u1(t)**2*sin(q2(t)) - l**2*m*Derivative(u2(t), t) - m*(l**2*cos(q2(t)) + l**2)*Derivative(u1(t), t)]])

In [28]: fr.simplify()

In [29]: fr
Out[29]: 
Matrix([
[-g*l*m*(sin(q1(t) + q2(t)) + 2*sin(q1(t)))],
[                 -g*l*m*sin(q1(t) + q2(t))]])

In [30]: frstar.simplify()

In [31]: frstar
Out[31]: 
Matrix([
[l**2*m*((u1(t) + u2(t))**2*sin(q2(t)) - (cos(q2(t)) + 1)*Derivative(u2(t), t) - (2*cos(q2(t)) + 3)*Derivative(u1(t), t) - u1(t)**2*sin(q2(t)))],
[                                 -l**2*m*(u1(t)**2*sin(q2(t)) + cos(q2(t))*Derivative(u1(t), t) + Derivative(u1(t), t) + Derivative(u2(t), t))]])

In [32]: M = kane.mass_matrix_full

In [33]: F = kane.forcing_full

In [34]: M.simplify()

In [35]: F.simplify()

In [36]: M
Out[36]: 
Matrix([
[1, 0,                         0,                       0],
[0, 1,                         0,                       0],
[0, 0, l**2*m*(2*cos(q2(t)) + 3), l**2*m*(cos(q2(t)) + 1)],
[0, 0,   l**2*m*(cos(q2(t)) + 1),                  l**2*m]])

In [37]: F
Out[37]: 
Matrix([
[                                                                                            u1(t)],
[                                                                                            u2(t)],
[l*m*(-g*sin(q1(t) + q2(t)) - 2*g*sin(q1(t)) + 2*l*u1(t)*u2(t)*sin(q2(t)) + l*u2(t)**2*sin(q2(t)))],
[                                              -l*m*(g*sin(q1(t) + q2(t)) + l*u1(t)**2*sin(q2(t)))]])
```

I think that is correct, but still double check against the wikipedia answer.

Reply:
Fixed, I was just missing forces.